streams/src/test/java/org/apache/kafka/streams/tests/EosTestDriver.java

```diff
@@ -559,47 +601,36 @@ private static void verifyAllTransactionFinished(final KafkaConsumer<byte[], byt
                         System.err.flush();
                         Exit.exit(1);
+                    } else {
+                        endMarkerOffset.put(new TopicPartition(metadata.topic(), metadata.partition()), metadata.offset());
+                        System.out.println("Appended verification record to topic-partition " + metadata.topic() + "-" + metadata.partition() + " at offset " + metadata.offset());
+                        System.out.flush();
                     }
                 });
             }
         }
 
-        final StringDeserializer stringDeserializer = new StringDeserializer();
-
         long maxWaitTime = System.currentTimeMillis() + MAX_IDLE_TIME_MS;
-        while (!partitions.isEmpty() && System.currentTimeMillis() < maxWaitTime) {
-            final ConsumerRecords<byte[], byte[]> records = consumer.poll(Duration.ofMillis(100));
-            if (records.isEmpty()) {
-                System.out.println("No data received.");
-                for (final TopicPartition tp : partitions) {
-                    System.out.println(tp + " at position " + consumer.position(tp));
-                }
-            }
-            for (final ConsumerRecord<byte[], byte[]> record : records) {
-                maxWaitTime = System.currentTimeMillis() + MAX_IDLE_TIME_MS;
-                final String topic = record.topic();
-                final TopicPartition tp = new TopicPartition(topic, record.partition());
+        while (!endMarkerOffset.isEmpty() && System.currentTimeMillis() < maxWaitTime) {
+            consumer.seekToEnd(partitions);
 
-                try {
-                    final String key = stringDeserializer.deserialize(topic, record.key());
-                    final String value = stringDeserializer.deserialize(topic, record.value());
+            final Iterator<TopicPartition> iterator = partitions.iterator();
+            while(iterator.hasNext()) {
+                final TopicPartition topicPartition = iterator.next();
 
-                    if (!("key".equals(key) && "value".equals(value) && partitions.remove(tp))) {
-                        throw new RuntimeException("Post transactions verification failed. Received unexpected verification record: " +
-                            "Expected record <'key','value'> from one of " + partitions + " but got"
-                            + " <" + key + "," + value + "> [" + record.topic() + ", " + record.partition() + "]");
-                    } else {
-                        System.out.println("Verifying " + tp + " successful.");
-                    }
-                } catch (final SerializationException e) {
-                    throw new RuntimeException("Post transactions verification failed. Received unexpected verification record: " +
-                        "Expected record <'key','value'> from one of " + partitions + " but got " + record, e);
+                if (consumer.position(topicPartition) > endMarkerOffset.get(topicPartition)) {
```

Comment:
Strictly, `position` should be exactly `endMarkerOffset + 1` -- it seems ok to just check for `>`

Reply:
Why we want to relax this check here?

Reply:
No reason. I can make it strict, too.